calico/felix/ipsets.py

```diff
@@ -528,32 +530,35 @@ def replace_members(self, members, force_reprogram=False):
         """
         Replace the members of this ipset with the supplied set.
 
-        :param set[str]|list[str] members: IP address strings.
+        :param set[str]|list[str] members: IP address strings. Must be a copy
+        (as this routine keeps a link to it).
         """
         _log.info("Replacing members of ipset %s", self.name)
-        self.members.clear()
-        self.members.update(members)
+        self.members = members
         self._force_reprogram |= force_reprogram
 
     def _finish_msg_batch(self, batch, results):
         _log.debug("IpsetActor._finish_msg_batch() called")
-        if not self.stopped and (self._force_reprogram or
-                                 self.members != self.programmed_members):
-            _log.debug("IpsetActor not in sync, updating dataplane.")
+        if not self.stopped:
             self._sync_to_ipset()
-            self._force_reprogram = False
 
     def _sync_to_ipset(self):
-        _log.debug("Setting ipset %s to %s", self, self.members)
         # Defer to our helper to actually make the changes.
-        if self._force_reprogram or self.programmed_members is None:
+        if self._force_reprogram:
+            _log.debug("Replacing content of ipset %s with %s", self,
+                       self.members)
             self._ipset.replace_members(self.members)
-            self.programmed_members = self.members.copy()
-        elif self.members != self.programmed_members:
+        elif self.programmed_members != self.members:
+            assert(self.programmed_members is not None)
```

Comment:
Nitpick, I think assert shouldn't have parens because they can be misleading.  In particular, `assert(False, "message")` looks right but it's actually equivalent to `x = (False, "message); assert x`, which is wrong.